File: app/src/main/java/top/easelink/lcg/ui/main/articles/view/FavoriteArticlesFragment.java
package top.easelink.lcg.ui.main.articles.view;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import top.easelink.framework.BR;
import top.easelink.framework.base.BaseFragment;
import top.easelink.lcg.R;
import top.easelink.lcg.databinding.FragmentFavoriteArticlesBinding;
import top.easelink.lcg.ui.ViewModelProviderFactory;
import top.easelink.lcg.ui.main.articles.viewmodel.FavoriteArticlesAdapter;
import top.easelink.lcg.ui.main.articles.viewmodel.FavoriteArticlesViewModel;

import javax.inject.Inject;

import static top.easelink.lcg.ui.main.articles.viewmodel.FavoriteArticlesViewModel.FETCH_INIT;

public class FavoriteArticlesFragment extends BaseFragment<FragmentFavoriteArticlesBinding, FavoriteArticlesViewModel>
        implements ArticlesNavigator {

    @Inject
    ViewModelProviderFactory factory;
    private FragmentFavoriteArticlesBinding mFragmentArticlesBinding;
    private FavoriteArticlesViewModel mArticlesViewModel;

    public static FavoriteArticlesFragment newInstance() {
        return new FavoriteArticlesFragment();
    }

    @Override
    public int getBindingVariable() {
        return BR.viewModel;
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_favorite_articles;
    }

    @Override
    public FavoriteArticlesViewModel getViewModel() {
        mArticlesViewModel = ViewModelProviders.of(this, factory).get(FavoriteArticlesViewModel.class);
        return mArticlesViewModel;
    }

    @Override
    public void handleError(Throwable t) {
        Toast.makeText(getContext(), R.string.error, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void scrollToTop() {
        mFragmentArticlesBinding.recyclerView.smoothScrollToPosition(0);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mArticlesViewModel.setNavigator(this);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mFragmentArticlesBinding = getViewDataBinding();
        setUpRecyclerView();
        setupMenu();
        mArticlesViewModel.fetchArticles(FETCH_INIT);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        menu.clear();
        inflater.inflate(R.menu.favorite_articles, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_remove_all:
                mArticlesViewModel.removeAllFavorites();
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void showMessage(@StringRes int resId) {
        Toast.makeText(getContext(), resId, Toast.LENGTH_SHORT).show();
    }

    private void setUpRecyclerView() {
        FavoriteArticlesAdapter mArticlesAdapter = new FavoriteArticlesAdapter(mArticlesViewModel);
        LinearLayoutManager mLayoutManager = new LinearLayoutManager(getContext());
        mLayoutManager.setOrientation(RecyclerView.VERTICAL);
        mFragmentArticlesBinding.recyclerView.setLayoutManager(mLayoutManager);
        mFragmentArticlesBinding.recyclerView.setItemAnimator(new DefaultItemAnimator());
        mFragmentArticlesBinding.recyclerView.setAdapter(mArticlesAdapter);
    }

    private void setupMenu() {
        setHasOptionsMenu(true);
        ((AppCompatActivity) getActivity()).setSupportActionBar(mFragmentArticlesBinding.articleToolbar);
    }
}

File: app/src/main/java/top/easelink/lcg/ui/main/articles/view/FavoriteArticlesFragment.kt
package top.easelink.lcg.ui.main.articles.view

import android.os.Bundle
import android.view.View
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.fragment_favorite_articles.*
import top.easelink.framework.base.BaseFragment
import top.easelink.lcg.BR
import top.easelink.lcg.R
import top.easelink.lcg.databinding.FragmentFavoriteArticlesBinding
import top.easelink.lcg.ui.main.articles.viewmodel.ArticleFetcher
import top.easelink.lcg.ui.main.articles.viewmodel.FavoriteArticlesViewModel
import top.easelink.lcg.ui.webview.view.WebViewActivity
import top.easelink.lcg.utils.WebsiteConstant.GET_FAVORITE_QUERY
import top.easelink.lcg.utils.WebsiteConstant.SERVER_BASE_URL

class FavoriteArticlesFragment : BaseFragment<FragmentFavoriteArticlesBinding, FavoriteArticlesViewModel>() {

    override fun isControllable(): Boolean {
        return true
    }

    override fun getBindingVariable(): Int {
        return BR.viewModel
    }

    override fun getLayoutId(): Int {
        return R.layout.fragment_favorite_articles
    }

    override fun getViewModel(): FavoriteArticlesViewModel {
        return ViewModelProvider(this)[FavoriteArticlesViewModel::class.java]
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?
    ) {
        super.onViewCreated(view, savedInstanceState)
        setUpRecyclerView()
        setupToolBar()
        viewModel.fetchArticles(ArticleFetcher.FetchType.FETCH_INIT){}
    }

    private fun setUpRecyclerView() {
        viewDataBinding.recyclerView.apply {
            layoutManager = LinearLayoutManager(context).apply {
                orientation = RecyclerView.VERTICAL
            }
            itemAnimator = DefaultItemAnimator()
            val adapter = FavoriteArticlesAdapter(viewModel)
            this.adapter = adapter
            ItemTouchHelper(ItemTouchHelperCallback(adapter)).attachToRecyclerView(this)
        }
        viewModel.articles.observe(viewLifecycleOwner, Observer {
            (viewDataBinding.recyclerView.adapter as? FavoriteArticlesAdapter)?.apply {
                clearItems()
                addItems(it)
            }
        })

    }

    private fun setupToolBar() {
        article_toolbar.apply {
            inflateMenu(R.menu.favorite_articles)
            setOnMenuItemClickListener {
                when (it.itemId) {
                    R.id.action_remove_all -> viewModel.removeAllFavorites()
                    R.id.action_sync_my_favorites -> viewModel.syncFavorites()
                    R.id.action_manage_favorites -> WebViewActivity.startWebViewWith(SERVER_BASE_URL + GET_FAVORITE_QUERY, context)
                    else -> {
                        // to add mores
                    }
                }
                return@setOnMenuItemClickListener true
            }
        }

    }
}